cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh

```diff
@@ -42,38 +42,60 @@ namespace detail {
  * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible and mutable.
  *
+ * @param[in] point_geometry_offset_first Iterator to the beginning of the range of the multipoint
+ * parts
+ * @param[in] point_geometry_offset_last Iterator to the end of the range of the multipoint parts
  * @param[in] points_first Iterator to the beginning of the range of the points
  * @param[in] points_last  Iterator to the end of the range of the points
- * @param[in] linestring_offsets_begin Iterator to the beginning of the range of the linestring
+ * @param[in] linestring_geometry_offset_first Iterator to the beginning of the range of the
+ * linestring parts
+ * @param[in] linestring_geometry_offset_last Iterator to the end of the range of the linestring
+ * parts
+ * @param[in] linestring_part_offsets_first Iterator to the beginning of the range of the linestring
  * offsets
- * @param[in] linestring_offsets_end Iterator to the end of the range of the linestring offsets
- * @param[in] linestring_points_begin Iterator to the beginning of the range of the linestring
+ * @param[in] linestring_part_offsets_last Iterator to the beginning of the range of the linestring
+ * offsets
+ * @param[in] linestring_points_first Iterator to the beginning of the range of the linestring
  * points
- * @param[in] linestring_points_end Iterator to the end of the range of the linestring points
+ * @param[in] linestring_points_last Iterator to the end of the range of the linestring points
  * @param[out] distances Iterator to the output range of shortest distances between pairs.
  *
  * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
  * "LegacyRandomAccessIterator"
  */
-template <typename Cart2dItA, typename Cart2dItB, typename OffsetIterator, typename OutputIterator>
-void __global__ pairwise_point_linestring_distance(Cart2dItA points_first,
-                                                   OffsetIterator linestring_offsets_first,
-                                                   OffsetIterator linestring_offsets_last,
+template <class Cart2dItA,
+          class Cart2dItB,
+          class OffsetIteratorA,
+          class OffsetIteratorB,
+          class OffsetIteratorC,
+          class OutputIterator>
+void __global__ pairwise_point_linestring_distance(OffsetIteratorA point_geometry_offset_first,
+                                                   OffsetIteratorA point_geometry_offset_last,
+                                                   Cart2dItA points_first,
+                                                   Cart2dItA points_last,
+                                                   OffsetIteratorB linestring_geometry_offset_first,
+                                                   OffsetIteratorB linestring_geometry_offset_last,
+                                                   OffsetIteratorC linestring_part_offsets_first,
+                                                   OffsetIteratorC linestring_part_offsets_last,
```

Comment:
I believe you are just looking at a recursive templated iterator collection, there's some maximum level that c++ compiler allows the templates to be recursively nested within itself, but multilinestring should be well within that limit. 

Does this make sense to you? https://godbolt.org/z/v95jn5G3e

Reply:
Yes, that's more or less exactly what I was thinking of. Do you think it works in the context of your PR?

Reply:
This really is just my first attempt to take a stab at the idea. I'd like to have some more feedbacks before we decide on the iterator collection interface. So no, not planning to put this in this PR.

cc @harrism

Reply:
There could be other simpler solutions, like just a `MultiLinestringIteratorCollection` class for example:
```c++
template<typename OffsetItA, typename OffsetItB, typename Vec2dIt>
struct MultiLinestringIteratorCollection
{
  OffsetItA geometry_offset_begin;
  OffsetItB part_offset_begin;
  Vec2dIt coordinate_begin;
};
```

Reply:
This deviates a little bit from you original suggestion of having a custom iterator to represent a nested of iterator. I believe you meant to create a custom iterator that iterates over the multilinestring array, when you dereference this iterator, it returns a multilinestring - which in turn has its own iterator, and allows you to iterate into each linestring, segment and points etc. Besides, we should provide different level of iterations so that developer can pick their favorite parallel model as needed.

I don't have a working example, but something like:
```c++
class linestring {
    class iterator{
        std::pair<vec2d, vec2d> operator* (); // returns two end points of the line segment
        iterator operator+(int i);
        iterator operator-(int i);
        // other requirements of of LRAIs
    };
    iterator begin();
    iterator end();
};

class multi_linestring
{
    class linestring_iterator
    {
        linestring operator*(); // this returns a linestring
        // LRAI requirements
    };
    class segment_iterator
    {
        std::pair<vec2d, vec2d> operator* (); // returns two end points of the line segment
        // LRAI requirments
    }

    linestring_iterator linestring_begin();
    linestring_iterator linestring_end();

    segment_iterator segment_begin();
    segment_iterator segment_end();

    // The default should still be immediate next level of indirection
    auto begin() { return linestring_begin(); }
    auto end() { return linestring_end(); }
};

class multi_linestring_array
{
    class multi_linestring_iterator
    {
        multi_linestring operator*();
        // other LRAI requirements
    }
    class linestring_iterator { /* same as above */ }
    class segment_iterator { /* same as above */ }

    multilinestring_iterator multilinestring_begin();
    multilinestring_iterator multilinestring_end();

    linestring_iterator linestring_begin();
    linestring_iterator linestring_end();

    segment_iterator segment_begin();
    segment_iterator segment_end();

    // The default should still be immediate next level of indirection
    auto begin() { return multilinestring_begin(); }
    auto end() { return multilinestring_begin(); }
};
```